pkg/storage/endpoint/gc_states.go

```diff
@@ -0,0 +1,526 @@
+		rounded := expirationTime.Truncate(time.Second)
+		if rounded.Before(*expirationTime) {
+			rounded = rounded.Add(time.Second)
+		}
+		*expirationTime = rounded
+	}
+	return &GCBarrier{
+		BarrierID:      barrierID,
+		BarrierTS:      barrierTS,
+		ExpirationTime: expirationTime,
+	}
+}
+
+// gcBarrierFromServiceSafePoint returns the GCBarrier that's synonymous to the given service safe point.
+func gcBarrierFromServiceSafePoint(s *ServiceSafePoint) *GCBarrier {
+	if s == nil {
+		return nil
+	}
+
+	res := &GCBarrier{
+		BarrierID:      s.ServiceID,
+		BarrierTS:      s.SafePoint,
+		ExpirationTime: nil,
+	}
+	if s.ExpiredAt < math.MaxInt64 && s.ExpiredAt > 0 {
+		expirationTime := new(time.Time)
+		*expirationTime = time.Unix(s.ExpiredAt, 0)
+		res.ExpirationTime = expirationTime
+	}
+	return res
+}
+
+// toServiceSafePoint converts the GCBarrier to a synonymous ServiceSafePoint for storing physically.
+func (b *GCBarrier) toServiceSafePoint(keyspaceID uint32) *ServiceSafePoint {
+	res := &ServiceSafePoint{
+		ServiceID:  b.BarrierID,
+		ExpiredAt:  math.MaxInt64,
+		SafePoint:  b.BarrierTS,
+		KeyspaceID: keyspaceID,
+	}
+	if b.ExpirationTime != nil {
+		res.ExpiredAt = b.ExpirationTime.Unix()
+	}
+	return res
+}
+
+// IsExpired checks whether the GCBarrier is expired at the given time.
+func (b *GCBarrier) IsExpired(now time.Time) bool {
+	return b.ExpirationTime != nil && now.After(*b.ExpirationTime)
+}
+
+// String implements fmt.Stringer.
+func (b *GCBarrier) String() string {
+	expirationTime := "<nil>"
+	if b.ExpirationTime != nil {
+		expirationTime = b.ExpirationTime.String()
+	}
+	return fmt.Sprintf("GCBarrier { BarrierID: %s, BarrierTS: %d, ExpirationTime: %v }",
+		b.BarrierID, b.BarrierTS, expirationTime)
+}
+
+// GCStateStorage is the interface for providing the ability to store and retrieve GC state data.
+// The GC state data is not available to access via the GCStateStorage interface; instead, it should be
+// accessed by the GCStateProvider indirectly, which can be retrieved by calling GetGCStateProvider.
+type GCStateStorage interface {
+	// GetGCStateProvider returns an GCStateProvider for reading and writing GC state data.
+	GetGCStateProvider() GCStateProvider
+}
+
+// GetGCStateProvider returns an GCStateProvider for reading and writing GC state data.
+// GCStateProvider provides direct read/write to GC related metadata without maintaining the consistency or invariants.
+// It's only designed for internal use of GC. Avoid using it for other purposes.
+func (se *StorageEndpoint) GetGCStateProvider() GCStateProvider {
+	return newGCStateProvider(se)
+}
+
+// GCStateProvider is a stateless wrapper over StorageEndpoint that provides methods for reading/writing GC states.
+// It can be dangerous to misuse GC related operations. As an explicit wrapper, it hides the GC related methods away
+// from the StorageEndpoint type and the Storage interface, making it less likely to be misused unintentionally when
```

Comment:
Do we need to ensure the order inside `GCStateProvider`? For example, check `GCBarrier >= TxnSafePoint` when we set GCBarrier or set TxnSafePoint?

Reply:
No.  As said in the comments at the beginning of the `gc_states.go` file, this layer is only responsible for providing methods for read and write. The upper layer `GCStateManager`, which will come in the next PR, is responsible for maintaining the invariants.